### PR TITLE
[stable25]  Add fallback routines for empty secret cases

### PR DESCRIPTION
**Pull request by CarlSchwan**
Fix https://github.com/nextcloud/server/issues/35347
Fix https://github.com/nextcloud/server/issues/35443

Manual backport of https://github.com/nextcloud/server/pull/31499

### CODE REVIEW

**Review by PVince81 — APPROVED** [review]
:+1:

File: lib/private/Authentication/Token/PublicKeyTokenProvider.php [hunk 1]
```diff
@@ -119,8 +119,14 @@ public function getToken(string $tokenId): IToken {
...
 				$this->cache[$token->getToken()] = $token;
 			} catch (DoesNotExistException $ex) {
-				$this->cache[$tokenHash] = $ex;
-				throw new InvalidTokenException("Token does not exist: " . $ex->getMessage(), 0, $ex);
+				try {
+					$token = $this->mapper->getToken($this->hashTokenWithEmptySecret($tokenId));
+					$this->cache[$token->getToken()] = $token;
+					$this->rotate($token, $tokenId, $tokenId);
+				} catch (DoesNotExistException $ex2) {
+					$this->cache[$tokenHash] = $ex2;
```
Comment (come-nc) [inline lib/private/Authentication/Token/PublicKeyTokenProvider.php:127]:
```suggestion
					$this->cache[$tokenHash] = $ex;
```

**Review by szaimen — APPROVED** [review]
🐘

**Comment by PVince81** [comment]
since this partly affects external storages, let's rerun the tests once now that https://github.com/nextcloud/server/pull/35606 is merged

**Pull request merged**